File: backend/app/api/endpoints/gelbooru.py
"""Gelbooru tag search proxy endpoint."""

from fastapi import APIRouter, HTTPException, Query, status

from app.api.deps import CurrentUser
from app.schemas.gelbooru import GelbooruTagSearchResponse
from app.services.gelbooru_service import (
    GelbooruRateLimitError,
    GelbooruUnavailableError,
    GelbooruUpstreamError,
    gelbooru_service,
)

router = APIRouter(prefix="/gelbooru", tags=["gelbooru"])


@router.get("/tags", response_model=GelbooruTagSearchResponse)
async def search_gelbooru_tags(
    _: CurrentUser,
    q: str = Query(..., min_length=2, description="Search query for tag name"),
    limit: int = Query(30, ge=1, le=100, description="Max number of results"),
) -> GelbooruTagSearchResponse:
    """Search Gelbooru tags by partial match.

    Proxies requests to Gelbooru API with caching.
    """
    try:
        tags = await gelbooru_service.search_tags(q, limit)
    except GelbooruRateLimitError as e:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail="Gelbooru API rate limit exceeded. Please try again later.",
        ) from e
    except GelbooruUnavailableError as e:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Gelbooru API is temporarily unavailable.",
        ) from e
    except GelbooruUpstreamError as e:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail="Gelbooru API returned an unexpected response.",
        ) from e
    return GelbooruTagSearchResponse(tags=tags, query=q)

File: backend-rs/src/http/gelbooru.rs
//! Gelbooru tag-search proxy handler (mirror endpoints/gelbooru.py).

use axum::extract::{Query, State};
use axum::Json;

use super::auth::CurrentUser;
use super::AppState;
use crate::dto::gelbooru::GelbooruTagSearchResponse;
use crate::error::AppError;
use crate::gelbooru::{self, GelbooruError};

#[derive(Debug, serde::Deserialize)]
pub struct TagQuery {
    pub q: String,
    #[serde(default = "default_limit")]
    pub limit: i64,
}

fn default_limit() -> i64 {
    30
}

/// GET /api/gelbooru/tags?q=&limit=30
pub async fn search_tags(
    _user: CurrentUser,
    State(state): State<AppState>,
    Query(query): Query<TagQuery>,
) -> Result<Json<GelbooruTagSearchResponse>, AppError> {
    if query.q.chars().count() < 2 {
        return Err(AppError::BadRequest(
            "q must be at least 2 characters".to_string(),
        ));
    }
    let limit = query.limit.clamp(1, 100);
    let tags = gelbooru::search_tags(
        &state.config.gelbooru_api_key,
        &state.config.gelbooru_user_id,
        &query.q,
        limit,
    )
    .await
    .map_err(|e| match e {
        GelbooruError::RateLimit => AppError::TooManyRequests(
            "Gelbooru API rate limit exceeded. Please try again later.".to_string(),
        ),
        GelbooruError::Unavailable => {
            AppError::ServiceUnavailable("Gelbooru API is temporarily unavailable.".to_string())
        }
        GelbooruError::Upstream => {
            AppError::BadGateway("Gelbooru API returned an unexpected response.".to_string())
        }
    })?;

    Ok(Json(GelbooruTagSearchResponse {
        tags,
        query: query.q,
    }))
}
